src/Storages/MergeTree/MergeTreeReaderStream.h

```diff
@@ -64,4 +65,34 @@ class MergeTreeReaderStream
 
     MergeTreeMarksLoader marks_loader;
 };
+
+/// A wrapper class over MergeTreeReaderStream that allows
+/// deferring opening the file until it's contents are requested
+class LazyMergeTreeReaderStream
```

Comment:
I suggest to make the `MergeTreeReaderStream` class lazy itself instead of creating this lazy wrapper:

```
class MergeTreeReaderStream
{
     ...
     bool initialized = false;
};

MergeTreeReaderStream::MergeTreeReaderStream(...)
{
    // The costructor is now empty
}

void MergeTreeReaderStream::init()
{
    if (initialized)
        return;
    initialized = true;
    
    /// Former constructor's body here
    ...
}

size_t MergeTreeReaderStream::getRightOffset(size_t right_mark_non_included)
{
    init();
    ...
}


void MergeTreeReaderStream::seekToMark(size_t index)
{
     init();
     ...
}
```

Reply:
Thank you for your feedback @vitlibar, I'll make the changes you suggest.

Reply:
Yes, thank you for noticing, of course `data_buffer` should be initialized before use